### ISSUE TITLE
warning CS0436: The type ‘SelfRegisteredExtensions’ causes a conflict

### ISSUE THREAD

**Issue by Kiwibomber** [xunit/xunit#3036]
warning CS0436: The type ‘SelfRegisteredExtensions’ in ‘...\bugtest\obj\Debug\net8.0\AutoRegisteredExtensions.cs’ causes a conflict with the imported type ‘SelfRegisteredExtensions’ in ‘ClassLibrary1, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null’. The type defined in ‘...\bugtest\obj\Debug\net8.0\AutoRegisteredExtensions.cs’ is used.

super fresh warning with the following version:
![image](https://github.com/user-attachments/assets/71999ca0-f68d-486b-a6f8-c2a7a79f031d)

and this is how you can recreate it:
one solution with two libs. the first lib has a ProjectReference to the second lib.
in the second lib exists the following class:
`    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = true)]
    public class BaseTestAttribute : Attribute
    {
        public BaseTestAttribute()
        {
        }
    }`

**Comment by bradwilson** [comment]
Yeah, this is an issue with Microsoft Testing Platform that's already being resolved.

This is technically [reported against MSTest](https://github.com/microsoft/testfx/issues/3813) but the [fix is in MTP](https://github.com/microsoft/testfx/pull/3891).

Once MTP 1.5 is shipped, we will take an updated dependency to resolve the issue.

**Comment by MarcoK80** [comment]
@bradwilson  This issue is not resolved if you have two projects and internalsvisible used to reuse an internal class.
Example: 
UnitTest Project A has a FakeImplementation class that you want to reuse. InternalsVisibleTo is set for Project B
UnitTest Project B wants to use the Fakeimplementation class of Project A.

**Comment by bradwilson** [comment]
@MarcoK80 It's their bug, not ours, so please make sure to comment on their PR/issue.

**Comment by ivan-luxoft** [comment]
Hi guys, I was met this issue too.
I have a few test projects (integration/ui) and one uses other (reuses some base methods). Both of them have tests, changing output type and another ways is not working for me, but just skipping warn allows to avoid this error (https://learn.microsoft.com/en-us/visualstudio/ide/how-to-suppress-compiler-warnings?view=vs-2022)
something like:
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <NoWarn>CS0436</NoWarn>
  </PropertyGroup>
...
</Project>
  ```
  
That's not a properly way but it's working, may will be useful for you too

**Comment by martincostello** [comment]
Just ran into this myself updating a xunit extension library of mine to support v3.

The linked PR into the test platform seems to be merged - is there a dependency somewhere that my project or xunit needs to pick up a newer version of to resolve the problem?

For now I've worked around things by adding the following property to the test project for my library:

```xml
<PropertyGroup>
  <GenerateSelfRegisteredExtensions>false</GenerateSelfRegisteredExtensions>
</PropertyGroup>
```

**Comment by bradwilson** [comment]
I think this was already covered in another issue, but to document it here as well: the fix is for extension libraries to take a dependency on `xunit.v3.extensibility.core`, not `xunit.v3` or `xunit.v3.core`.

Thanks!

**Comment by bradwilson** [comment]
> I believe that "making a test project to depend on another test project" is not what people should do

Correct, unit test projects should not reference unit test projects.

If you have code you want to share between two unit test projects, put it in a third non-unit test project. If it needs access to xUnit.net types, the non-test project can reference `xunit.v3.extensibility.core` and/or `xunit.v3.assert` as needed.

> Was it ever a good idea?

Probably not. Despite the fact that unit test projects have historically been built as library projects, they have always effectively been applications of their own, but with the "entry point" being the test runner. Now that v3 has made this formalized (and the "entry point" is actually a Main function that we inject), consider this much like the situation where one executable references another executable.

**Comment by bradwilson** [comment]
Closing as "external".

**Comment by bradwilson** [comment]
Pulling in a new version of `Microsoft.Testing.Platform.MSBuild` will be done once they've shipped and then we ship a new version.

That said, you wouldn't actually need to wait for us, because MTPM is only expressed as a NuGet dependency:

https://github.com/xunit/xunit/blob/a3876f4883c15db418bacf3c37b287eacf6d7856/src/xunit.v3.core/xunit.v3.core.nuspec#L19

Which means you can always explicit choose to take a newer version as a dependency yourself.

**Comment by nabeelio** [comment]
I'm getting this after updating to xunit3, but my test projects don't depend on one another. Just droppin that in here - I'm looking at the upstream tickets for resolution
